tx-pool/src/component/container.rs

```diff
@@ -161,118 +134,90 @@ impl SortedTxMap {
             let parent_hash = &input.previous_output().tx_hash();
             let id = ProposalShortId::from_tx_hash(&(parent_hash));
-            if self.links.contains_key(&id) {
+            if self.contains_key(&id) {
                 parents.insert(id);
             }
         }
         for cell_dep in entry.transaction.cell_deps() {
             let id = ProposalShortId::from_tx_hash(&(cell_dep.out_point().tx_hash()));
-            if self.links.contains_key(&id) {
+            if self.contains_key(&id) {
                 parents.insert(id);
             }
         }
         // update ancestor_fields
         self.update_ancestors_stat_for_entry(&mut entry, &parents);
         // update parents references
         for parent_id in &parents {
-            self.links
-                .get_mut(parent_id)
+            self.get_link_mut(parent_id)
                 .expect("exists")
                 .children
                 .insert(short_id.clone());
         }
-        // insert links
-        self.links.insert(
-            short_id.clone(),
-            TxLink {
-                parents,
-                children: Default::default(),
-            },
-        );
-        self.sorted_index
-            .insert(AncestorsScoreSortKey::from(&entry));
-        self.entries.insert(short_id, entry)
-    }
-
-    pub fn contains_key(&self, id: &ProposalShortId) -> bool {
-        self.entries.contains_key(id)
-    }
 
-    pub fn get(&self, id: &ProposalShortId) -> Option<&TxEntry> {
-        self.entries.get(id)
+        let link = TxLink {
+            parents,
+            children: Default::default(),
+        };
+        self.insert(entry, link)
     }
 
     pub fn remove_entry_and_descendants(&mut self, id: &ProposalShortId) -> Vec<TxEntry> {
-        let mut queue = VecDeque::new();
-        let mut removed = Vec::new();
-        queue.push_back(id.clone());
-        while let Some(id) = queue.pop_front() {
-            if let Some(entry) = self.entries.remove(&id) {
-                let deleted = self
-                    .sorted_index
-                    .remove(&AncestorsScoreSortKey::from(&entry));
-                debug_assert!(deleted, "pending pool inconsistent");
-                if let Some(link) = self.links.remove(&id) {
-                    queue.extend(link.children);
-                }
-                removed.push(entry);
-            }
+        let mut removal = Vec::new();
+
+        if let Some(entry) = self.get(id) {
+            removal.push(entry.to_owned());
+        } else {
+            return removal;
         }
-        removed
+        let descendants = self.get_descendants(id);
+        removal.extend(descendants.iter().map(|desc_id| {
+            self.get(desc_id)
+                .expect("found inconsistency when remove_entry_and_descendants")
+                .to_owned()
+        }));
+
+        removal.iter().for_each(|e| {
+            self.remove_entry(&e.transaction.proposal_short_id());
```

Comment:
Obviously, this implementation has `log(N * N)` complexity compare with the old `log(N)`, please don't do this.